File: app/src/main/java/com/amaury/pointage/v2/engine/FullTimeStructuralOvertimeV2.kt
package com.amaury.pointage.v2.engine

/** Mensualisation d'un temps plein dont la durée contractuelle peut dépasser la durée légale/référence. */
object FullTimeStructuralOvertimeV2 {
    data class TierAmount(val multiplier:Double,val minutes:Double,val gross:Double)
    data class Result(
        val monthlyBaseGross:Double,
        val monthlyRegularMinutes:Double,
        val monthlyStructuralOvertimeMinutes:Double,
        val structuralOvertimeGross:Double,
        val variableOvertimeGross:Double,
        val structuralTiers:List<TierAmount>,
        val variableTiers:List<TierAmount>,
        /**
         * Compatibilité API : true signifie désormais qu'au moins une tranche n'a pas pu être
         * valorisée faute de taux confirmé. Aucun taux de secours n'est injecté dans le montant.
         */
        val provisionalRateUsed:Boolean,
        val unresolvedStructuralOvertimeMinutes:Double,
        val unresolvedVariableOvertimeMinutes:Double,
        val warnings:List<String>
    )

    fun calculate(
        contractualWeeklyMinutes:Int,
        regularWeeklyLimit:Int,
        paidWeeks:List<Int>,
        grossHourlyRate:Double,
        overtimeTiers:List<OvertimeTierV2>
    ):Result {
        require(contractualWeeklyMinutes>0)
        require(regularWeeklyLimit>0)
        require(grossHourlyRate>0.0)

        // Réutilise la même barrière canonique que PayrollEngineV2 : un jeu de paliers ambigu
        // ou invalide ne doit jamais alimenter directement un montant, quelle que soit la plateforme.
        val tiersStructurallyValid=OvertimeCoverageV2.isStructurallyValid(regularWeeklyLimit,overtimeTiers)
        val safeOvertimeTiers=OvertimeCoverageV2.calculationSafeTiers(regularWeeklyLimit,overtimeTiers)

        val factor=52.0/12.0
        val regularContractMinutes=minOf(contractualWeeklyMinutes,regularWeeklyLimit)
        val structuralWeekly=ratedBetween(
            upper=contractualWeeklyMinutes,
            lower=regularWeeklyLimit,
            rate=grossHourlyRate,
            tiers=safeOvertimeTiers
        )
        val monthlyRegularMinutes=regularContractMinutes*factor
        val monthlyStructuralMinutes=structuralWeekly.minutes*factor
        val structuralGrossMonthly=structuralWeekly.gross*factor
        val monthlyBaseGross=monthlyRegularMinutes/60.0*grossHourlyRate+structuralGrossMonthly

        val variableParts=paidWeeks.map { paid ->
            ratedBetween(
                upper=paid.coerceAtLeast(0),
                lower=maxOf(contractualWeeklyMinutes,regularWeeklyLimit),
                rate=grossHourlyRate,
                tiers=safeOvertimeTiers
            )
        }
        val variableGross=variableParts.sumOf{it.gross}
        val allRated=listOf(structuralWeekly)+variableParts
        val warnings=buildList {
            addAll(allRated.flatMap{it.warnings})
            if(!tiersStructurallyValid&&overtimeTiers.isNotEmpty()) {
                add("Paliers d'heures supplémentaires ambigus ou invalides : ils sont neutralisés et aucune valorisation n'est inventée pour les tranches concernées.")
            }
        }.distinct()

        fun aggregate(parts:List<Rated>,monthly:Boolean):List<TierAmount> = parts
            .flatMap{it.tiers}
            .groupBy{it.multiplier}
            .map{(multiplier,items)->
                val minutes=items.sumOf{it.minutes}*(if(monthly)factor else 1.0)
                TierAmount(multiplier,minutes,items.sumOf{it.gross}*(if(monthly)factor else 1.0))
            }
            .sortedBy{it.multiplier}

        return Result(
            monthlyBaseGross=monthlyBaseGross,
            monthlyRegularMinutes=monthlyRegularMinutes,
            monthlyStructuralOvertimeMinutes=monthlyStructuralMinutes,
            structuralOvertimeGross=structuralGrossMonthly,
            variableOvertimeGross=variableGross,
            structuralTiers=aggregate(listOf(structuralWeekly),true),
            variableTiers=aggregate(variableParts,false),
            provisionalRateUsed=allRated.any{it.provisionalRateUsed},
            unresolvedStructuralOvertimeMinutes=structuralWeekly.unresolvedMinutes*factor,
            unresolvedVariableOvertimeMinutes=variableParts.sumOf{it.unresolvedMinutes},
            warnings=warnings
        )
    }

    private data class Piece(val multiplier:Double,val minutes:Double,val gross:Double)
    private data class Rated(
        val minutes:Double,
        val gross:Double,
        val tiers:List<Piece>,
        val provisionalRateUsed:Boolean,
        val unresolvedMinutes:Double,
        val warnings:List<String>
    )

    /**
     * Analyse la tranche (lower, upper] sans jamais laisser disparaître une minute.
     * Une minute sans palier confirmé reste comptée mais n'alimente aucun montant.
     */
    private fun ratedBetween(
        upper:Int,
        lower:Int,
        rate:Double,
        tiers:List<OvertimeTierV2>
    ):Rated {
        if(upper<=lower)return Rated(0.0,0.0,emptyList(),false,0.0,emptyList())
        val sorted=tiers.sortedBy{it.fromMinutes}
        var cursor=lower
        var gross=0.0
        var unresolvedMinutes=0.0
        val pieces=mutableListOf<Piece>()
        val warnings=mutableListOf<String>()

        fun add(from:Int,to:Int,multiplier:Double){
            if(to<=from)return
            val minutes=(to-from).toDouble()
            val amount=minutes/60.0*rate*multiplier
            pieces+=Piece(multiplier,minutes,amount)
            gross+=amount
        }

        fun addUnresolved(from:Int,to:Int){
            if(to<=from)return
            unresolvedMinutes+=(to-from).toDouble()
            warnings+="Palier d'heures supplémentaires non confirmé : aucune valorisation n'est injectée pour les minutes non couvertes ; le brut reste à confirmer."
        }

        sorted.forEach{tier->
            if(cursor>=upper)return@forEach
            val tierStart=maxOf(lower,tier.fromMinutes)
            val tierEnd=minOf(upper,tier.toMinutes?:Int.MAX_VALUE)
            if(tierEnd<=cursor||tierEnd<=tierStart)return@forEach
            if(tierStart>cursor){
                addUnresolved(cursor,minOf(tierStart,upper))
                cursor=minOf(tierStart,upper)
            }
            if(cursor<upper&&tierEnd>cursor){
                add(cursor,tierEnd,tier.multiplier)
                cursor=tierEnd
            }
        }
        if(cursor<upper)addUnresolved(cursor,upper)
        return Rated(
            minutes=(upper-lower).toDouble(),
            gross=gross,
            tiers=pieces,
            provisionalRateUsed=unresolvedMinutes>0.0,
            unresolvedMinutes=unresolvedMinutes,
            warnings=warnings.distinct()
        )
    }
}

File: ios/HPTravail/HPTravail/SalaryV2/FullTimeStructuralOvertimeV2.swift
import Foundation

/// Miroir Swift de la mensualisation Android d'un temps plein dont la durée contractuelle
/// peut dépasser la durée hebdomadaire régulière confirmée.
enum FullTimeStructuralOvertimeV2 {
    struct TierAmount: Equatable {
        let multiplier: Double
        let minutes: Double
        let gross: Double
    }

    struct Result: Equatable {
        let monthlyBaseGross: Double
        let monthlyRegularMinutes: Double
        let monthlyStructuralOvertimeMinutes: Double
        let structuralOvertimeGross: Double
        let variableOvertimeGross: Double
        let structuralTiers: [TierAmount]
        let variableTiers: [TierAmount]
        /// Compatibilité API : true signifie qu'au moins une tranche n'a pas pu être
        /// valorisée faute de taux confirmé. Aucun taux de secours n'est injecté.
        let provisionalRateUsed: Bool
        let unresolvedStructuralOvertimeMinutes: Double
        let unresolvedVariableOvertimeMinutes: Double
        let warnings: [String]
    }

    private struct Piece {
        let multiplier: Double
        let minutes: Double
        let gross: Double
    }

    private struct Rated {
        let minutes: Double
        let gross: Double
        let tiers: [Piece]
        let provisionalRateUsed: Bool
        let unresolvedMinutes: Double
        let warnings: [String]
    }

    static func calculate(
        contractualWeeklyMinutes: Int,
        regularWeeklyLimit: Int,
        paidWeeks: [Int],
        grossHourlyRate: Double,
        overtimeTiers: [OvertimeTierV2]
    ) -> Result {
        precondition(contractualWeeklyMinutes > 0)
        precondition(regularWeeklyLimit > 0)
        precondition(grossHourlyRate > 0 && grossHourlyRate.isFinite)

        let tiersStructurallyValid = OvertimeCoverageV2.isStructurallyValid(
            regularLimitMinutes: regularWeeklyLimit,
            tiers: overtimeTiers
        )
        let safeOvertimeTiers = OvertimeCoverageV2.calculationSafeTiers(
            regularLimitMinutes: regularWeeklyLimit,
            tiers: overtimeTiers
        )

        let factor = 52.0 / 12.0
        let regularContractMinutes = min(contractualWeeklyMinutes, regularWeeklyLimit)
        let structuralWeekly = ratedBetween(
            upper: contractualWeeklyMinutes,
            lower: regularWeeklyLimit,
            rate: grossHourlyRate,
            tiers: safeOvertimeTiers
        )
        let monthlyRegularMinutes = Double(regularContractMinutes) * factor
        let monthlyStructuralMinutes = structuralWeekly.minutes * factor
        let structuralGrossMonthly = structuralWeekly.gross * factor
        let monthlyBaseGross = monthlyRegularMinutes / 60.0 * grossHourlyRate + structuralGrossMonthly

        let variableParts = paidWeeks.map { paid in
            ratedBetween(
                upper: max(0, paid),
                lower: max(contractualWeeklyMinutes, regularWeeklyLimit),
                rate: grossHourlyRate,
                tiers: safeOvertimeTiers
            )
        }
        let variableGross = variableParts.reduce(0.0) { $0 + $1.gross }
        let allRated = [structuralWeekly] + variableParts
        var warnings = unique(allRated.flatMap(\.warnings))
        if !tiersStructurallyValid && !overtimeTiers.isEmpty {
            warnings.append("Paliers d'heures supplémentaires ambigus ou invalides : ils sont neutralisés et aucune valorisation n'est inventée pour les tranches concernées.")
        }

        return Result(
            monthlyBaseGross: monthlyBaseGross,
            monthlyRegularMinutes: monthlyRegularMinutes,
            monthlyStructuralOvertimeMinutes: monthlyStructuralMinutes,
            structuralOvertimeGross: structuralGrossMonthly,
            variableOvertimeGross: variableGross,
            structuralTiers: aggregate([structuralWeekly], monthly: true, factor: factor),
            variableTiers: aggregate(variableParts, monthly: false, factor: factor),
            provisionalRateUsed: allRated.contains { $0.provisionalRateUsed },
            unresolvedStructuralOvertimeMinutes: structuralWeekly.unresolvedMinutes * factor,
            unresolvedVariableOvertimeMinutes: variableParts.reduce(0.0) { $0 + $1.unresolvedMinutes },
            warnings: unique(warnings)
        )
    }

    /// Analyse la tranche (lower, upper] sans jamais laisser disparaître une minute.
    /// Une minute sans palier confirmé reste comptée mais n'alimente aucun montant.
    private static func ratedBetween(
        upper: Int,
        lower: Int,
        rate: Double,
        tiers: [OvertimeTierV2]
    ) -> Rated {
        if upper <= lower {
            return Rated(
                minutes: 0,
                gross: 0,
                tiers: [],
                provisionalRateUsed: false,
                unresolvedMinutes: 0,
                warnings: []
            )
        }

        let sorted = tiers.sorted { $0.fromMinutes < $1.fromMinutes }
        var cursor = lower
        var gross = 0.0
        var unresolvedMinutes = 0.0
        var pieces: [Piece] = []
        var warnings: [String] = []

        func piece(from: Int, to: Int, multiplier: Double) -> Piece? {
            guard to > from else { return nil }
            let minutes = Double(to - from)
            return Piece(
                multiplier: multiplier,
                minutes: minutes,
                gross: minutes / 60.0 * rate * multiplier
            )
        }

        func add(_ from: Int, _ to: Int, _ multiplier: Double) {
            guard let value = piece(from: from, to: to, multiplier: multiplier) else { return }
            pieces.append(value)
            gross += value.gross
        }

        func addUnresolved(_ from: Int, _ to: Int) {
            guard to > from else { return }
            unresolvedMinutes += Double(to - from)
            warnings.append("Palier d'heures supplémentaires non confirmé : aucune valorisation n'est injectée pour les minutes non couvertes ; le brut reste à confirmer.")
        }

        for tier in sorted {
            if cursor >= upper { continue }
            let tierStart = max(lower, tier.fromMinutes)
            let tierEnd = min(upper, tier.toMinutes ?? Int.max)
            if tierEnd <= cursor || tierEnd <= tierStart { continue }
            if tierStart > cursor {
                addUnresolved(cursor, min(tierStart, upper))
                cursor = min(tierStart, upper)
            }
            if cursor < upper && tierEnd > cursor {
                add(cursor, tierEnd, tier.multiplier)
                cursor = tierEnd
            }
        }
        if cursor < upper {
            addUnresolved(cursor, upper)
        }

        return Rated(
            minutes: Double(upper - lower),
            gross: gross,
            tiers: pieces,
            provisionalRateUsed: unresolvedMinutes > 0,
            unresolvedMinutes: unresolvedMinutes,
            warnings: unique(warnings)
        )
    }

    private static func aggregate(_ parts: [Rated], monthly: Bool, factor: Double) -> [TierAmount] {
        let scale = monthly ? factor : 1.0
        let grouped = Dictionary(grouping: parts.flatMap(\.tiers), by: \.multiplier)
        return grouped.map { multiplier, items in
            TierAmount(
                multiplier: multiplier,
                minutes: items.reduce(0.0) { $0 + $1.minutes } * scale,
                gross: items.reduce(0.0) { $0 + $1.gross } * scale
            )
        }.sorted { $0.multiplier < $1.multiplier }
    }

    private static func unique(_ values: [String]) -> [String] {
        var seen = Set<String>()
        return values.filter { seen.insert($0).inserted }
    }
}
